test/app-luatest/http_client_test.lua

```diff
@@ -1352,3 +1353,36 @@ g.test_http_client_io_post_parts = function(cg)
     t.assert_equals(io.status, 200, 'io')
     t.assert_equals(io.reason, 'Ok', '200 - Ok')
 end
+
+-- Read from ph:stderr and fail if it doesn't contain a `pattern'.
+local function grep_stderr_or_fail(ph, pattern)
+    local output = ''
+    t.helpers.retrying({}, function()
+        local chunk = ph:read({timeout = 0.05, stderr = true})
+        if chunk ~= nil then
+           output = output .. chunk
+        end
+        t.assert_str_contains(output, pattern)
+    end)
+end
+
+g.test_http_client_gc_finalizer_gh_9453 = function(cg)
+    local tarantool_exe = arg[-1]
+    local lua_code = [[local url = '%s'
+                       local c = require('http.client').new()
+                       local r = c:get(url, {chunked = true})
+
+                       c = nil
+                       collectgarbage()
+                       collectgarbage()
+
+                       r:read(1)]]
+    local ph = popen.new({ tarantool_exe, '-e', lua_code:format(cg.url) },
+                         { stderr = popen.opts.PIPE })
+    t.assert(ph)
+
+    -- Wait for the process startup.
+    grep_stderr_or_fail(ph, "IllegalParams: io: request must be io")
```

Comment:
> It's quite strange to add a test for "wrong" behaviour.

Okay, I can squash commits. For me, it is not strange because in this way I show old behavior and with fix change an assent. With squashed commits, all things will not be so obvious. 

> It's a bit ugly... Is it possible to implement it without popen in commit that fixes an issue?

No, with a repro inside a luatest testcase bug #9346 is triggered.